[setup] Ubuntu Jammy may have libgcc-12-dev, which breaks mkdoc if libstdc++-12-dev is not present

### What happened?

Relates https://github.com/RobotLocomotion/drake/issues/15652#issuecomment-904774598

See more details in https://drakedevelopers.slack.com/archives/C0KDGF4V9/p1669663654774429
\cc @calderpg-tri @ggould-tri @jwnimmer-tri @rpoyner-tri 

As Grant indicated, can repro as follows:
``
bazel build @drake//bindings/pydrake:documentation_pybind.h
``

If you have `libgcc-12-dev` installed, but *not* `libstdc++-12-dev`, then it will fail with something like:
```
bazel-out/k8-opt/bin/external/drake/bindings/pydrake/documentation_pybind.h.tmp_artifacts/mkdoc_glue.h:1:10: fatal error: 'optional' file not found
Ubuntu clang version 12.0.1-19ubuntu3'
...
RuntimeError: The operating system's C++ standard library is not installed correctly.
```

If you run off of bare Ubuntu install (e.g. `ubuntu:22.04` in Docker / Apptainer), then it's fine.
If you install `libgcc-12-dev`, then it breaks.
If you then install `libstdc++-12-dev`, then it's fine.

Also note that we configure `.bazelrc` as follows for Jammy on Anzu:
```sh
# Use GCC 11 to match Jammy's default.
build --action_env=CC=/usr/bin/gcc-11
build --action_env=CXX=/usr/bin/g++-11
```

### Version

2b49d9bf3a971620ec00673949d91339f8162c2c

### What operating system are you using?

Ubuntu 22.04

### What installation option are you using?

compiled from source code using Bazel

### Relevant log output

_No response_

--- a/setup/ubuntu/source_distribution/install_prereqs.sh
+++ b/setup/ubuntu/source_distribution/install_prereqs.sh
@@ -120,6 +120,21 @@ if [[ "${with_maintainer_only}" -eq 1 ]]; then
   apt-get install ${maybe_yes} --no-install-recommends ${packages}
 fi
 
+# On Jammy, Drake doesn't install anything related to GCC 12, but if the user
+# has chosen to install some GCC 12 libraries but has failed to install all of
+# them correctly as a group, Drake's documentation header file parser will fail
+# with a libclang-related complaint. Therefore, we'll help the user clean up
+# their mess, to avoid apparent Drake build errors.
+if [[ "${codename}" == "jammy" ]]; then
+  status=$(dpkg-query --show --showformat='${db:Status-Abbrev}' libgcc-12-dev 2>/dev/null || true)
+  if [[ "${status}" == "ii " ]]; then
+    status=$(dpkg-query --show --showformat='${db:Status-Abbrev}' libstdc++-12-dev 2>/dev/null || true)
+    if [[ "${status}" != "ii " ]]; then
+      apt-get install ${maybe_yes} --no-install-recommends libgcc-12-dev libstdc++-12-dev
+    fi
+  fi
+fi
+
 dpkg_install_from_wget() {
   package="$1"
   version="$2"

--- a/tools/workspace/pybind11/mkdoc.py
+++ b/tools/workspace/pybind11/mkdoc.py
@@ -824,8 +824,12 @@ def main():
                 except Exception:
                     pass
                 raise RuntimeError(
-                    ("The operating system's C++ standard library is not "
-                     "installed correctly."))
+                    "The operating system's C++ standard library is not "
+                    "installed correctly or is only partially installed. For "
+                    "example, libgcc-??-dev is installed but libstdc++-??-dev "
+                    "is not installed (the ?? indicates a version number). "
+                    "Try re-running Drake's install_prereqs, or maybe check "
+                    "your system and install anything that's missing by hand.")
         severities = [
             diagnostic.severity for diagnostic in translation_unit.diagnostics
             if diagnostic.severity >= cindex.Diagnostic.Error

